akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -155,14 +176,13 @@ object ByteString {
 
     def isCompact: Boolean = (length == bytes.length)
 
+    override def canWrapAsByteBuffer: Boolean = true
+
     def compact: CompactByteString =
       if (isCompact) ByteString1C(bytes) else ByteString1C(toArray)
 
-    def asByteBuffer: ByteBuffer = {
-      val buffer = ByteBuffer.wrap(bytes, startIndex, length).asReadOnlyBuffer
-      if (buffer.remaining < bytes.length) buffer.slice
```

Comment:
logic proved in test to be equivalent to isCompact. If expecting a compact ByteBuffer, one can simply ByteString#compact.asByteBuffer